baseplate/clients/thrift.py

```diff
@@ -213,38 +218,45 @@ def _call_thrift_method(self: Any, *args: Any, **kwargs: Any) -> Any:
                         prot.trans.set_header(b"Edge-Request", edge_context)
 
                     result = method(*args, **kwargs)
-            except TTransportException as exc:
-                # the connection failed for some reason, retry if able
-                span.finish(exc_info=sys.exc_info())
-                last_error = str(exc)
-                if exc.inner is not None:
-                    last_error += f" ({exc.inner})"
-                continue
-            except (TApplicationException, TProtocolException):
-                # these are subclasses of TException but aren't ones that
-                # should be expected in the protocol. this is an error!
-                span.finish(exc_info=sys.exc_info())
-                raise
-            except Error as exc:
-                # a 5xx error is an unexpected exception but not 5xx are
-                # not.
-                if 500 <= exc.code < 600:
+                except TTransportException as exc:
+                    # the connection failed for some reason, retry if able
+                    span.finish(exc_info=sys.exc_info())
+                    last_error = str(exc)
+                    if exc.inner is not None:
+                        last_error += f" ({exc.inner})"
+                    continue
+                except (TApplicationException, TProtocolException):
+                    # these are subclasses of TException but aren't ones that
+                    # should be expected in the protocol. this is an error!
                     span.finish(exc_info=sys.exc_info())
+                    raise
+                except Error as exc:
+                    # a 5xx error is an unexpected exception but not 5xx are
+                    # not.
+                    if 500 <= exc.code < 600:
+                        span.finish(exc_info=sys.exc_info())
+                    else:
+                        span.finish()
+                    raise
+                except TException:
+                    # this is an expected exception, as defined in the IDL
+                    span.finish()
+                    raise
+                except:  # noqa: E722
+                    # something unexpected happened
+                    try:
```

Comment:
Are we worried that something outside BP is catching this error so ParentSpanAlreadyFinishedError is not reported to sentry?

Reply:
This changes the behaviour, causing expected exceptions to get missed:

Old behavior:
```
☸ dev-3 (ws-kris-thomt) in baseplate.py on  k/prom-sql-client-metrics [$?] via 🐍 v3.9.13 (alt-text)
✦3 ❯ cat test.py
class SuperBadError(Exception):
    pass

class NotSoBadError(Exception):
    pass

try:
    try:
        raise SuperBadError
    except:
        raise
except SuperBadError:
    print("I am caught!")


☸ dev-3 (ws-kris-thomt) in baseplate.py on  k/prom-sql-client-metrics [$?] via 🐍 v3.9.13 (alt-text)
✦3 ❯ python3 test.py
I am caught!
```

New behavior:
```
☸ dev-3 (ws-kris-thomt) in baseplate.py on  k/prom-sql-client-metrics [$?] via 🐍 v3.9.13 (alt-text)
✦3 ❯ cat test.py
class SuperBadError(Exception):
    pass

class NotSoBadError(Exception):
    pass

try:
    try:
        raise SuperBadError
    except:
        try:
            raise NotSoBadError
        except SuperBadError:
            raise
        raise
except SuperBadError:
    print("I am caught")


☸ dev-3 (ws-kris-thomt) in baseplate.py on  k/prom-sql-client-metrics [$?] via 🐍 v3.9.13 (alt-text)
✦3 ❯ python3 test.py
Traceback (most recent call last):
  File "/home/krav/src/alt-text/baseplate.py/test.py", line 9, in <module>
    raise SuperBadError
__main__.SuperBadError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/krav/src/alt-text/baseplate.py/test.py", line 12, in <module>
    raise NotSoBadError
__main__.NotSoBadError
```

Reply:
It's not needed, since the exception chain `SomethingError<ParentSpanAlreadyFinishedError` is different from `SomethingError`. The user must explicitly catch `ParentSpanAlreadyFinishedError` for python not to error, `catch SomethingError` is not enough.